# Proteomining-Clustering
Proteomining clustering algorithm for biosynthetic gene cluster identification

Usage: python 'Proteomining Clustering.py' input proteindb genegap min-members
To run example:
python 'Proteomining Clustering.py' Example-data-set.csv 'S. coelicolor A3(2) Genome (GeneMarkS).fasta' 5 8

Arguments:

Input = excel or csv file (see below)

proteindb = fasta or faa files. Fasta files should be results of gene prediction with original ORF number preserved.

Optional:

genegap: Maximum allowed gaps between ORFs, Default = 5

min-members: Minimum nodes in a cluster to be considered, Default = 8



Outputs generated are the interactive html. 

Accessible by locally hosting:
python -m http.server 9000
Access in a browser at "localhost:9000"

This script accepts csv or EXCEL files as input. Filters and assigns ORFs as nodes that are differentially expressed.
Nodes are clustered based on the gap in ORF (based on assigned number generated by gene prediction).
ORF generated by GeneMarkS is preferred due to formatting preferences especially in partial genomes.
Result outputs in EXCEL format and as .html and json files.


Expected format for input:

ORF,log2(foldchange),-log10(p-value)

Each ORF is tagged as a node and are assigned scores:

log2(Fold-change) >= 1, 	-log10(p-value) >= 2 		: 10

log2(Fold-change) >= 0.584963, 	-log10(p-value) >= 1.30103 	: 5

log2(Fold-change) <= -0.584963, -log10(p-value) >= 1.30103 	: -5

log2(Fold-change) <= -1, 	-log10(p-value) >= 2 		: -10
